Plugin Extensions: Add prop types to component extensions

leventebalogh opened the pull request:
### What changed?

- Updated the `getPluginComponentExtensions()` function to receive the component prop types as a generic _(this makes it easier to define extension points with proper typing)_
- Removed the `context` option from the same function _(The contextual information can be passed to the component as props, so we decided to remove this additional option to make the setup less confusing.)_

```ts
// Before
// ---------------------------------
const { extensions } = getPluginComponentExtensions({ extensionPointId });

({ extensions.map(extension => {
  const Component = extension.component as React.ComponentType<{
    foo: string;
  }>;
})})


// After
// ---------------------------------
const { extensions } = getPluginComponentExtensions<{ foo: string }>({ extensionPointId });
```

leventebalogh committed:
feat: make it possible to specify prop types for component extensions

darrenjaneczek reviewed (approved):
:+1:

darrenjaneczek commented on packages/grafana-runtime/src/services/pluginExtensions/getPluginExtensions.ts:
```diff
@@ -41,10 +41,15 @@ export const getPluginLinkExtensions: GetPluginExtensions<PluginExtensionLink> =
 };
 
-export const getPluginComponentExtensions: GetPluginExtensions<PluginExtensionComponent> = (options) => {
+// This getter doesn't support the `context` option (contextual information can be passed in as component props) 
```
```suggestion
// This getter doesn't support the `context` option (contextual information can be passed in as component props)
```

whitespace

darrenjaneczek committed:
Update packages/grafana-runtime/src/services/pluginExtensions/getPluginExtensions.ts

darrenjaneczek committed:
chore: adapted test case

darrenjaneczek reviewed (changes requested):
Additional type declaration work may need to be done on `configureExtensionComponent`.

```ts
  configureExtensionComponent<Context extends object>(
    extension: Omit<PluginExtensionComponentConfig<Context>, 'type'>
  ) {
    this._extensionConfigs.push({
      ...extension,
      type: PluginExtensionTypes.component,
    } as PluginExtensionComponentConfig);

    return this;
  }
```

and/or the `PluginExtensionComponentConfig` type:

```ts
export type PluginExtensionComponentConfig<Context extends object = object> = {
  type: PluginExtensionTypes.component;
  title: string;
  description: string;

  // The React component that will be rendered as the extension
  // (This component receives the context as a prop when it is rendered. You can just return `null` from the component to hide for certain contexts)
  component: React.ComponentType<{
    context?: Context;
  }>;

  // The unique identifier of the Extension Point
  // (Core Grafana extension point ids are available in the `PluginExtensionPoints` enum)
  extensionPointId: string;
};
```

I think something like changing `Context` into `Props`, and getting rid of the `{context?: Context}` layer in the component field's type.

With the exception of this typescript error on `component` on the call `plugin.configureExtensionComponent<ComponentExtensionsProps> (...)`, I can verify that everything is working as expected with this no context approach.

leventebalogh commented:
Very good point, thanks for testing it out @darrenjaneczek 🙌
I will have a look at it and get the typing fixed.

leventebalogh committed:
chore: update betterer

leventebalogh committed:
feat: update types for configureComponentExtension()

darrenjaneczek committed:
fix: remove type specifics for `configureExtensionComponent`

darrenjaneczek reviewed (approved):
The type errors go away after a slight additional change:
- 37e502e

It does trigger a 'levitate breaking change' however...

leventebalogh committed:
Update betterer config

leventebalogh commented:
Thanks @darrenjaneczek, I think that fix is legit 👏 I have updated the betterer configs, hopefully they will be fine now.

**Levitate breaking change**
 It's complaining because we have changed the type of the generic for `configureExtensionComponent()`, however I don't think it should break anything, as the props objects should satisfy the previous type declaration of the generic (`object`).

mckn reviewed (approved):
LGTM!
